lib/jsonapi_compliable/util/hooks.rb

```diff
@@ -0,0 +1,30 @@
+module JsonapiCompliable
+  module Util
+    class Hooks
+      def self.record
+        self.hooks = []
+        begin
+          yield.tap { run }
+        ensure
+          self.hooks = []
+        end
+      end
+
+      def self._hooks
+        Thread.current[:_compliable_hooks] ||= []
+      end
+
+      def self.hooks=(val)
+        Thread.current[:_compliable_hooks] = val
+      end
+
+      def self.add(prc)
+        self._hooks.unshift(prc)
```

Comment:
Because we process the edges of the graph first, and work inwards. But the hooks should fire in the opposite order. This is tested [here](https://github.com/richmolj/jsonapi_compliable/blob/hooks_w/spec/integration/rails/before_commit_spec.rb#L212) and I added a comment about it.